src/libmongoc/src/mongoc/mongoc-topology-background-monitoring.c

```diff
@@ -0,0 +1,838 @@
+{
+   mongoc_topology_t *topology;
+
+   topology = topology_void;
+   bson_mutex_lock (&topology->mutex);
+   while (true) {
+      int64_t now_ms;
+      int64_t scan_due_ms;
+      int64_t sleep_duration_ms;
+
+      if (topology->scanner_state != MONGOC_TOPOLOGY_SCANNER_BG_RUNNING) {
+         bson_mutex_unlock (&topology->mutex);
+         break;
+      }
+
+      /* This will check if a scan is due. */
+      mongoc_topology_rescan_srv (topology);
+
+      /* Unlock and sleep until next scan is due, or until shutdown signalled.
+       */
+      now_ms = bson_get_monotonic_time () / 1000;
+      scan_due_ms = topology->srv_polling_last_scan_ms +
+                    topology->srv_polling_rescan_interval_ms;
+      sleep_duration_ms = scan_due_ms - now_ms;
+
+      if (sleep_duration_ms > 0) {
+         TRACE ("srv polling thread sleeping for %" PRId64 "ms",
+                sleep_duration_ms);
+      }
+
+      /* If shutting down, stop. */
+      mongoc_cond_timedwait (
+         &topology->srv_polling_cond, &topology->mutex, sleep_duration_ms);
+   }
+   BSON_THREAD_RETURN;
+}
+
+typedef struct {
+   mongoc_topology_t *topology;
+   bson_thread_t thread;
+
+   /* State accessed from multiple threads. */
+   struct {
+      bson_mutex_t mutex;
+      mongoc_cond_t cond;
+      bool shutting_down;
+      bool is_shutdown;
+      bool scan_requested;
+   } shared;
+
+   /* Time of last scan in milliseconds. */
+   uint64_t last_scan_ms;
+   /* The time of the next scheduled scan. */
+   uint64_t scan_due_ms;
+   /* The server id matching the server description. */
+   uint32_t server_id;
+   /* Default time to sleep between ismaster checks (reduced when a scan is
+    * requested) */
+   uint64_t heartbeat_frequency_ms;
+   /* The minimum time to sleep between ismaster checks. */
+   uint64_t min_heartbeat_frequency_ms;
+   int64_t connect_timeout_ms;
+   bool use_tls;
+#ifdef MONGOC_ENABLE_SSL
+   mongoc_ssl_opt_t *ssl_opts;
+#endif
+   mongoc_uri_t *uri;
+   mongoc_host_list_t host;
+   /* A custom initiator may be set if a user provides overrides to create a
+    * stream. */
+   mongoc_stream_initiator_t initiator;
+   void *initiator_context;
+   mongoc_stream_t *stream;
+   int64_t request_id;
+   mongoc_apm_callbacks_t apm_callbacks;
+   void *apm_context;
+} mongoc_server_monitor_t;
+
+/* Called only from server monitor thread.
+ * Caller must hold no locks (user's callback may lock topology mutex).
+ * Locks APM mutex.
+ */
+static void
+_server_monitor_heartbeat_started (mongoc_server_monitor_t *server_monitor)
+{
+   if (server_monitor->apm_callbacks.server_heartbeat_started) {
+      mongoc_apm_server_heartbeat_started_t event;
+      event.host = &server_monitor->host;
+      event.context = server_monitor->apm_context;
+      bson_mutex_lock (&server_monitor->topology->apm_mutex);
+      server_monitor->apm_callbacks.server_heartbeat_started (&event);
+      bson_mutex_unlock (&server_monitor->topology->apm_mutex);
+   }
```

Comment:
Nit: using a guard clause helps reduce nesting, but I'm aware that most code here wraps logic in the condition instead of exiting early.

```suggestion
   if (!server_monitor->apm_callbacks.server_heartbeat_started) {
      return;
   }
   
   mongoc_apm_server_heartbeat_started_t event;
   event.host = &server_monitor->host;
   event.context = server_monitor->apm_context;
   bson_mutex_lock (&server_monitor->topology->apm_mutex);
   server_monitor->apm_callbacks.server_heartbeat_started (&event);
   bson_mutex_unlock (&server_monitor->topology->apm_mutex);
```

Reply:
Good idea, done.